sideboard/lib/_profiler.py

```diff
@@ -0,0 +1,269 @@
+    See Also:
+        configspec.ini
+    """
+    if config['cherrypy'].get('profiling.on', False):
+        profiling_path = config['cherrypy'].get('profiling.path', None)
+        if config['cherrypy'].get('profiling.aggregate', False):
+            p = ProfileAggregator(profiling_path)
+        else:
+            p = Profiler(profiling_path)
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            return p.run(func, *args, **kwargs)
+        return wrapper
+    else:
+        return func
+
+
+class Profiler(object):
+    """
+    Mostly copied from cherrypy/lib/profiler.py.
+
+     * Adds the ability to sort results by different columns.
+     * Adds the ability to cleanup profile data files.
+     * Uses a better naming scheme for profile data files.
+    """
+
+    # https://docs.python.org/3/library/profile.html#pstats.Stats.sort_stats
+    sort_fields = [
+        ('cumulative', 'Cumulative Time'),
+        ('filename', 'File Name'),
+        ('ncalls', 'Call Count'),
+        ('pcalls', 'Primitive Call Count'),
+        ('line', 'Line Number'),
+        ('name', 'Function Name'),
+        ('nfl', 'Function/File/Line'),
+        ('stdname', 'Standard Name'),
+        ('tottime', 'Total Time')]
+
+    def __init__(self, path=None):
+        if not path:
+            path = os.path.join(os.path.dirname(__file__), 'profile')
+        self.path = path
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+    def new_filename(self, func):
+        date = datetime.now().strftime("%Y-%m-%d_%H:%M:%S.%f")
+        name = func.__name__ if func.__name__ else 'unknown'
+        return '{}_{}.prof'.format(date, name)
+
+    def run(self, func, *args, **params):
+        """Dump profile data into self.path."""
+        path = os.path.join(self.path, self.new_filename(func))
+        prof = cProfile.Profile()
+        result = prof.runcall(func, *args, **params)
+        prof.dump_stats(path)
+        return result
+
+    def statfiles(self):
+        """:rtype: list of available profiles.
+        """
+        return [f for f in os.listdir(self.path) if f.endswith('.prof')]
+
+    def stats(self, filename, sortby='cumulative'):
+        """:rtype stats(index): output of print_stats() for the given profile.
+        """
+        sio = io.StringIO()
+        s = pstats.Stats(os.path.join(self.path, filename), stream=sio)
+        if config['cherrypy'].get('profiling.strip_dirs', False):
+            s.strip_dirs()
+        s.sort_stats(sortby)
+        s.print_stats()
+        response = sio.getvalue()
+        sio.close()
+        return response
+
+    @cherrypy.expose
+    def index(self):
+        return '''<html>
+        <head><title>Sideboard Profiler</title></head>
+        <frameset cols="300, 1*">
+            <frame src="menu"/>
+            <frame name="main" src="" />
+        </frameset>
+        </html>
+        '''
+
+    @cherrypy.expose
+    def menu(self):
+        yield '<h2>Profiling Runs</h2>'
```

Comment:
I am not crazy about manually concatenating a bunch of HTML strings together, but this is how the cherrypy module was doing it, and I didn't feel like rewriting a bunch of working code.